File: trunk/arrg/ua_controllers/ax12_controller_core/src/ax12_controller_core/joint_position_controller.py
#!/usr/bin/env python
#
# Software License Agreement (BSD License)
#
# Copyright (c) 2010, Arizona Robotics Research Group,
# University of Arizona. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of University of Arizona nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# Author: Antons Rebguns
# Author: Cara Slutter
#

from __future__ import division

import roslib
roslib.load_manifest('ax12_controller_core')

import rospy
from ax12_driver_core.ax12_const import *
from ax12_driver_core.ax12_user_commands import *
from ax12_controller_core.joint_controller import JointControllerAX12
from ua_controller_msgs.msg import JointState

class JointPositionControllerAX12(JointControllerAX12):
    def __init__(self, out_cb, param_path, port_name):
        JointControllerAX12.__init__(self, out_cb, param_path, port_name)
        
        self.motor_id = rospy.get_param(self.topic_name + '/motor/id')
        self.initial_position_raw = rospy.get_param(self.topic_name + '/motor/init')
        self.min_angle_raw = rospy.get_param(self.topic_name + '/motor/min')
        self.max_angle_raw = rospy.get_param(self.topic_name + '/motor/max')
        
        self.flipped = self.min_angle_raw > self.max_angle_raw
        
        self.joint_state = JointState(name=self.joint_name, motor_ids=[self.motor_id])

    def initialize(self):
        # verify that the expected motor is connected and responding
        available_ids = rospy.get_param('dynamixel/%s/connected_ids' % self.port_namespace, [])
        if not self.motor_id in available_ids:
            rospy.logwarn('The specified motor id is not connected and responding.')
            rospy.logwarn('Available ids: %s' % str(available_ids))
            rospy.logwarn('Specified id: %d' % self.motor_id)
            return False
            
        self.radians_per_encoder_tick = rospy.get_param('dynamixel/%s/%d/radians_per_encoder_tick' % (self.port_namespace, self.motor_id))
        self.encoder_ticks_per_radian = rospy.get_param('dynamixel/%s/%d/encoder_ticks_per_radian' % (self.port_namespace, self.motor_id))
        
        if self.flipped:
            self.min_angle = (self.initial_position_raw - self.min_angle_raw) * self.radians_per_encoder_tick
            self.max_angle = (self.initial_position_raw - self.max_angle_raw) * self.radians_per_encoder_tick
        else:
            self.min_angle = (self.min_angle_raw - self.initial_position_raw) * self.radians_per_encoder_tick
            self.max_angle = (self.max_angle_raw - self.initial_position_raw) * self.radians_per_encoder_tick
            
        self.encoder_resolution = rospy.get_param('dynamixel/%s/%d/encoder_resolution' % (self.port_namespace, self.motor_id))
        self.max_position = self.encoder_resolution - 1
        self.set_speed(self.joint_speed)
        
        if self.compliance_slope is not None: self.set_compliance_slope(self.compliance_slope)
        if self.compliance_margin is not None: self.set_compliance_margin(self.compliance_margin)
        if self.compliance_punch is not None: self.set_compliance_punch(self.compliance_punch)
        if self.torque_limit is not None: self.set_torque_limit(self.torque_limit)
        return True

    def set_torque_enable(self, torque_enable):
        mcv = (self.motor_id, torque_enable)
        self.send_packet_callback((DMXL_SET_TORQUE_ENABLE, [mcv]))

    def set_speed(self, speed):
        if speed < DMXL_MIN_SPEED_RAD: speed = DMXL_MIN_SPEED_RAD
        elif speed > self.joint_max_speed: speed = self.joint_max_speed
        speed_raw = int(round(speed / DMXL_SPEED_RAD_SEC_PER_TICK))
        mcv = (self.motor_id, speed_raw)
        self.send_packet_callback((DMXL_SET_GOAL_SPEED, [mcv]))

    def set_compliance_slope(self, slope):
        if slope < DMXL_MIN_COMPLIANCE_SLOPE: slope = DMXL_MIN_COMPLIANCE_SLOPE
        elif slope > DMXL_MAX_COMPLIANCE_SLOPE: slope = DMXL_MAX_COMPLIANCE_SLOPE
        slope2 = (slope << 8) + slope
        mcv = (self.motor_id, slope2)
        self.send_packet_callback((DMXL_SET_COMPLIANCE_SLOPES, [mcv]))

    def set_compliance_margin(self, margin):
        if margin < DMXL_MIN_COMPLIANCE_MARGIN: margin = DMXL_MIN_COMPLIANCE_MARGIN
        elif margin > DMXL_MAX_COMPLIANCE_MARGIN: margin = DMXL_MAX_COMPLIANCE_MARGIN
        else: margin = int(margin)
        margin2 = (margin << 8) + margin    # pack margin_cw and margin_ccw into 2 bytes
        mcv = (self.motor_id, margin2)
        self.send_packet_callback((DMXL_SET_COMPLIANCE_MARGINS, [mcv]))

    def set_compliance_punch(self, punch):
        if punch < DMXL_MIN_PUNCH: punch = DMXL_MIN_PUNCH
        elif punch > DMXL_MAX_PUNCH: punch = DMXL_MAX_PUNCH
        else: punch = int(punch)
        mcv = (self.motor_id, punch)
        self.send_packet_callback((DMXL_SET_PUNCH, [mcv]))

    def set_torque_limit(self, max_torque):
        if max_torque > 1: max_torque = 1.0         # use all torque motor can provide
        elif max_torque < 0: max_torque = 0.0       # turn off motor torque
        raw_torque_val = int(DMXL_MAX_TORQUE_TICK * max_torque)
        mcv = (self.motor_id, raw_torque_val)
        self.send_packet_callback((DMXL_SET_TORQUE_LIMIT, [mcv]))

    def process_motor_states(self, state_list):
        if self.running:
            state = filter(lambda state: state.id == self.motor_id, state_list.motor_states)
            if state:
                state = state[0]
                self.joint_state.motor_temps = [state.temperature]
                self.joint_state.goal_pos = self.raw_to_rad(state.goal, self.initial_position_raw, self.flipped, self.radians_per_encoder_tick)
                self.joint_state.current_pos = self.raw_to_rad(state.position, self.initial_position_raw, self.flipped, self.radians_per_encoder_tick)
                self.joint_state.error = state.error * self.radians_per_encoder_tick
                self.joint_state.velocity = (state.speed / DMXL_MAX_SPEED_TICK) * DMXL_MAX_SPEED_RAD
                self.joint_state.load = state.load
                self.joint_state.is_moving = state.moving
                self.joint_state.header.stamp = rospy.Time.from_sec(state.timestamp)
                
                self.joint_state_pub.publish(self.joint_state)

    def process_command(self, msg):
        angle = msg.data
        if angle < self.min_angle: angle = self.min_angle
        elif angle > self.max_angle: angle = self.max_angle
        mcv = (self.motor_id, self.rad_to_raw(angle, self.initial_position_raw, self.flipped, self.encoder_ticks_per_radian))
        self.send_packet_callback((DMXL_SET_GOAL_POSITION, [mcv]))

File: arrg/ua_experimental/dynamixel_hardware_interface/src/joint_position_controller.cpp
/*
    Copyright (c) 2011, Antons Rebguns <email>
    All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are met:
        * Redistributions of source code must retain the above copyright
        notice, this list of conditions and the following disclaimer.
        * Redistributions in binary form must reproduce the above copyright
        notice, this list of conditions and the following disclaimer in the
        documentation and/or other materials provided with the distribution.
        * Neither the name of the <organization> nor the
        names of its contributors may be used to endorse or promote products
        derived from this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY Antons Rebguns <email> ''AS IS'' AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
    DISCLAIMED. IN NO EVENT SHALL Antons Rebguns <email> BE LIABLE FOR ANY
    DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
    (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
    LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
    ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include <sstream>
#include <XmlRpcValue.h>

#include <dynamixel_hardware_interface/dynamixel_const.h>
#include <dynamixel_hardware_interface/dynamixel_io.h>
#include <dynamixel_hardware_interface/single_joint_controller.h>
#include <dynamixel_hardware_interface/joint_position_controller.h>
#include <dynamixel_hardware_interface/MotorState.h>
#include <dynamixel_hardware_interface/SetVelocity.h>
#include <dynamixel_hardware_interface/TorqueEnable.h>

#include <ros/ros.h>
#include <pluginlib/class_list_macros.h>

#include <std_msgs/Float64.h>

PLUGINLIB_DECLARE_CLASS(dynamixel_hardware_interface,
                        JointPositionController,
                        controller::JointPositionController,
                        controller::SingleJointController)

namespace controller
{

bool JointPositionController::initialize(std::string name,
                                         std::string port_namespace,
                                         dynamixel_hardware_interface::DynamixelIO* dxl_io)
{
    if (!SingleJointController::initialize(name, port_namespace, dxl_io)) { return false; }

    c_nh_.getParam("motor/id", motor_id_);

    joint_state_.name = joint_;
    joint_state_.motor_ids.push_back(motor_id_);
    joint_state_.motor_temps.resize(1);

    std::stringstream ss;
    ss << "dynamixel/" << port_namespace_ << "/";
    
    std::string prefix = ss.str();
    XmlRpc::XmlRpcValue available_ids;
    
    nh_.getParam(prefix + "connected_ids", available_ids);
    
    if (available_ids.getType() != XmlRpc::XmlRpcValue::TypeArray)
    {
        ROS_ERROR("%s: [%s] paramater is not an array", (prefix + "connected_ids").c_str(), name_.c_str());
        return false;
    }
    
    bool found_motor_id = false;
    
    for (int i = 0; i < available_ids.size(); ++i)
    {
        XmlRpc::XmlRpcValue val = available_ids[i];
        if (motor_id_ == (int) val) { found_motor_id = true; break; }
    }
    
    if (!found_motor_id)
    {
        ROS_ERROR("%s: motor id %d is not connected to port %s or it is connected and not responding",
                  name_.c_str(), motor_id_, port_namespace_.c_str());
        return false;
    }
    
    ss  << motor_id_ << "/";
    prefix = ss.str();
    
    /***************** Joint velocity related parameters **********************/
    nh_.getParam(prefix + "max_velocity", motor_max_velocity_);
    nh_.getParam(prefix + "radians_second_per_encoder_tick", velocity_per_encoder_tick_);
    min_velocity_ = velocity_per_encoder_tick_;
    
    if (max_velocity_ > motor_max_velocity_)
    {
        ROS_WARN("%s: requested maximum joint velocity exceeds motor capabilities (%f > %f)", name_.c_str(), max_velocity_, motor_max_velocity_);
        max_velocity_ = motor_max_velocity_;
    }
    else if (max_velocity_ < min_velocity_)
    {
        ROS_WARN("%s: requested maximum joint velocity is too small (%f < %f)", name_.c_str(), max_velocity_, min_velocity_);
        max_velocity_ = min_velocity_;
    }

    motor_cache_ = dxl_io_->getCachedParameters(motor_id_);
    if (motor_cache_ == NULL) { return false; }

    int encoder_resolution;
    nh_.getParam(prefix + "encoder_resolution", encoder_resolution);
    
    if (motor_cache_->cw_angle_limit == 0 && motor_cache_->ccw_angle_limit == 0)
    {
        ROS_WARN("%s: motor %d is not set to position control mode, setting motor to position control mode", name_.c_str(), motor_id_);
        
        if (!dxl_io_->setAngleLimits(motor_id_, 0, encoder_resolution-1))
        {
            ROS_ERROR("%s: unable to set motor to position control mode", name_.c_str());
            return false;
        }
    }

    /***************** Motor related parameters **********************/
    c_nh_.getParam("motor/init", init_position_encoder_);
    c_nh_.getParam("motor/min", min_angle_encoder_);
    c_nh_.getParam("motor/max", max_angle_encoder_);
    
    flipped_ = min_angle_encoder_ > max_angle_encoder_;
    
    nh_.getParam(prefix + "radians_per_encoder_tick", radians_per_encoder_tick_);
    nh_.getParam(prefix + "encoder_ticks_per_radian", encoder_ticks_per_radian_);
    
    if (flipped_)
    {
        min_angle_radians_ = (init_position_encoder_ - min_angle_encoder_) * radians_per_encoder_tick_;
        max_angle_radians_ = (init_position_encoder_ - max_angle_encoder_) * radians_per_encoder_tick_;
    }
    else
    {
        min_angle_radians_ = (min_angle_encoder_ - init_position_encoder_) * radians_per_encoder_tick_;
        max_angle_radians_ = (max_angle_encoder_ - init_position_encoder_) * radians_per_encoder_tick_;
    }

    setVelocity(INIT_VELOCITY);
    return true;
}

void JointPositionController::start()
{
    SingleJointController::start();
}

void JointPositionController::stop()
{
    SingleJointController::stop();
}

std::vector<int> JointPositionController::getMotorIDs()
{
    return std::vector<int>(1, motor_id_);
}

std::vector<std::vector<int> > JointPositionController::getRawMotorCommands(double position, double velocity)
{
    std::vector<std::vector<int> > value_pairs;

    std::vector<int> pair;
    pair.push_back(motor_id_);
    pair.push_back(posRad2Enc(position));
    pair.push_back(velRad2Enc(velocity));
    
    value_pairs.push_back(pair);

    return value_pairs;
}

void JointPositionController::setVelocity(double velocity)
{
    std::vector<int> pair;
    pair.push_back(motor_id_);
    pair.push_back(velRad2Enc(velocity));
    
    std::vector<std::vector<int> > mcv;
    mcv.push_back(pair);

    dxl_io_->setMultiVelocity(mcv);
}

void JointPositionController::processMotorStates(const dynamixel_hardware_interface::MotorStateListConstPtr& msg)
{
    dynamixel_hardware_interface::MotorState state;
    
    for (size_t i = 0; i < msg->motor_states.size(); ++i)
    {
        if (motor_id_ == msg->motor_states[i].id)
        {
            state = msg->motor_states[i];
            break;
        }
    }
    
    if (state.id != motor_id_)
    {
        ROS_ERROR("%s: motor id %d not found", name_.c_str(), motor_id_);
        return;
    }
    
    joint_state_.header.stamp = ros::Time(state.timestamp);
    joint_state_.motor_temps[0] = state.temperature;
    joint_state_.target_position = convertToRadians(state.target_position);
    joint_state_.target_velocity = ((double)state.target_velocity / dynamixel_hardware_interface::DXL_MAX_VELOCITY_ENCODER) * motor_max_velocity_;
    joint_state_.position = convertToRadians(state.position);
    joint_state_.velocity = ((double)state.velocity / dynamixel_hardware_interface::DXL_MAX_VELOCITY_ENCODER) * motor_max_velocity_;
    joint_state_.load = (double)state.load / dynamixel_hardware_interface::DXL_MAX_LOAD_ENCODER;
    joint_state_.moving = state.moving;
    
    joint_state_pub_.publish(joint_state_);
}

void JointPositionController::processCommand(const std_msgs::Float64ConstPtr& msg)
{
    std::vector<int> pair;
    pair.push_back(motor_id_);
    pair.push_back(posRad2Enc(msg->data));
    
    std::vector<std::vector<int> > mcv;
    mcv.push_back(pair);
    
    dxl_io_->setMultiPosition(mcv);
}

bool JointPositionController::processSetVelocity(dynamixel_hardware_interface::SetVelocity::Request& req,
                                                 dynamixel_hardware_interface::SetVelocity::Request& res)
{
    setVelocity(req.velocity);
    return true;
}

bool JointPositionController::processTorqueEnable(dynamixel_hardware_interface::TorqueEnable::Request& req,
                                                  dynamixel_hardware_interface::TorqueEnable::Request& res)
{
    std::vector<int> pair;
    pair.push_back(motor_id_);
    pair.push_back(req.torque_enable);
    
    std::vector<std::vector<int> > mcv;
    mcv.push_back(pair);
    
    dxl_io_->setMultiTorqueEnabled(mcv);
    return true;
}

uint16_t JointPositionController::posRad2Enc(double pos_rad)
{
    if (pos_rad < min_angle_radians_) { pos_rad = min_angle_radians_; }
    if (pos_rad > max_angle_radians_) { pos_rad = max_angle_radians_; }
    return convertToEncoder(pos_rad);
}

uint16_t JointPositionController::velRad2Enc(double vel_rad)
{
    if (vel_rad < min_velocity_) { vel_rad = min_velocity_; }
    if (vel_rad > max_velocity_) { vel_rad = max_velocity_; }
    
    return std::max<uint16_t>(1, (uint16_t) round(vel_rad / velocity_per_encoder_tick_));
}

}
